Add support for load_data from URI/URL

bmorris3 opened the pull request:
This PR implements parser support for `viz.load_data(uri)` where `uri` is either a MAST URI or a web URL, for all `*viz` other than Mosviz.

Download and cache a remote FITS image on load into Imviz:
```python
from jdaviz import Imviz

url = "https://www.astropy.org/astropy-data/tutorials/FITS-images/HorseHead.fits"

imviz = Imviz()
imviz.load_data(url, cache=True)
imviz.show()
```
Generic URLs are downloaded with `astropy.utils.data.download_file`

Download a FITS spectrum from MAST on load to Specviz: 
```python
from jdaviz import Specviz

uri = "mast:JWST/product/jw02732-o004_t004_miri_ch1-shortmediumlong_x1d.fits"

specviz = Specviz()
specviz.load_data(uri)
specviz.show()
```
MAST queries are done via `astroquery.mast.Observations`. A Cubeviz example:
```python
from jdaviz import Cubeviz

uri = "mast:JWST/product/jw01373-o031_t007_miri_ch1-shortmediumlong_s3d.fits"

cubeviz = Cubeviz()
cubeviz.load_data(uri)
cubeviz.show()
```
and a Specviz2d example:
```python
from jdaviz import Specviz2d

uri = "mast:JWST/product/jw01324-o006_s00005_nirspec_f100lp-g140h_s2d.fits"

specviz2d = Specviz2d()
specviz2d.load_data(uri)
specviz2d.show()
```

cc @havok2063.

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-534)

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +390,52 @@ def total_masked_first_data(self):
+
+def download_uri_to_path(possible_uri, cache=False, local_path=None):
+    """
+    Retrieve data from a URI (or a URL). Return the input if it
```
I think this needs to explain that depending on input content, it could go through `astroquery.mast` or `astropy.utils.download_file`, which are two very different beasts. The former `cache` is only a `bool` but the latter isn't (bool or "update", see https://docs.astropy.org/en/stable/api/astropy.utils.data.download_file.html#astropy.utils.data.download_file).

Also should maybe say that local file URI (`file:\\`) is not supported.

And what about Cloud FITS (https://docs.astropy.org/en/stable/io/fits/usage/cloud.html#subsetting-fits-files-hosted-in-amazon-s3-cloud-storage) that has `s3:\\`?

pllim replied:
I also worry about URI injection but hopefully we can delegate that upstream.

bmorris3 replied:
I've added notes about this to the docstring. As for S3 – we will need to handle this for Roman, and likely with some Roman-specific optimizations. For now, I'm going to leave these excluded from the URI parser, though they might still work if you use the astropy functions to get a streaming HDU.

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +390,52 @@ def total_masked_first_data(self):
         return cls(masks=masks)
+
+
+def download_uri_to_path(possible_uri, cache=False, local_path=None):
```
Maybe default should be to always use cache, if available?
```suggestion
def download_uri_to_path(possible_uri, cache=True, local_path=None):
```

bmorris3 replied:
I prefer to keep it false by default – for users who cache a lot, it can be surprising and really frustrating to have significant disk space used by files without human readable names, making it hard to clean up without deleting all.

pllim replied:
But the other side is that you keep downloading the same thing over and over again, which can get expensive via AWS.

bmorris3 replied:
If our files were guaranteed to be small, I wouldn't blink. But I have cached a bunch of 5 GB NIRCam images, and regret that I did.

pllim replied:
I also have regretted accidentally re-downloading GB of data I already have. So, I don't think there is a perfect solution.

Now I think we should have a user doc about this new feature and document all the gotchas there.

rosteen replied:
I don't have strong feelings either way, but here's a third option: what if, since there are good arguments both ways, we require the user to explicitly set the `cache` argument in the case of loading from a URL? We could default to `cache=None` rather than `True` or `False` and then check to see if it's `None` in the appropriate case and raise an error telling the user to set it explicitly.

pllim replied:
@rosteen , but you still need a default for "standalone app", do you not?

rosteen replied:
Hmm yes that's true 😭

pllim commented on jdaviz/configs/cubeviz/plugins/parsers.py:
```diff
@@ -23,7 +23,7 @@
 
 @data_parser_registry("cubeviz-data-parser")
-def parse_data(app, file_obj, data_type=None, data_label=None, parent=None):
+def parse_data(app, file_obj, data_type=None, data_label=None, parent=None, cache=True):
```
Same comment as https://github.com/spacetelescope/jdaviz/pull/2875/files#r1602284711

pllim replied:
Ooops nvm.

pllim commented on jdaviz/configs/imviz/plugins/parsers.py:
```diff
@@ -43,7 +43,7 @@ def prep_data_layer_as_dq(data):
 
 @data_parser_registry("imviz-data-parser")
-def parse_data(app, file_obj, ext=None, data_label=None, parent=None):
+def parse_data(app, file_obj, ext=None, data_label=None, parent=None, cache=False):
```
Same comment as https://github.com/spacetelescope/jdaviz/pull/2875/files#r1602284711

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -9,15 +9,15 @@
 @data_parser_registry("specviz-spectrum1d-parser")
 def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_viewer=True,
-                              concat_by_file=False):
+                              concat_by_file=False, cache=False):
```
Same comment as https://github.com/spacetelescope/jdaviz/pull/2875/files#r1602284711

rosteen commented on jdaviz/configs/cubeviz/plugins/parsers.py:
```diff
@@ -66,6 +66,9 @@ def parse_data(app, file_obj, data_type=None, data_label=None, parent=None):
             return
 
+        # try parsing file_obj as a URI/URL:
+        file_obj = download_uri_to_path(file_obj, cache=cache)
```
This seems to not be handling the case of the user passing the path to a local file on disk correctly:

```
File [~/projects/jdaviz/jdaviz/utils.py:452](http://localhost:8888/lab/tree/~/projects/jdaviz/jdaviz/utils.py#line=451), in download_uri_to_path(possible_uri, cache, local_path)
    449     return download_file(possible_uri, cache=cache)
    451 else:
--> 452     raise ValueError(f"URI {possible_uri} with scheme {parsed_uri.scheme} is not "
    453                      f"currently supported.")
    455 # assume this isn't a URI after all:
    456 return possible_uri

ValueError: URI [/User/username/Data/test_cubes/jw02732-o004_t004_miri_ch1-shortmediumlong_s3d.fits](http://localhost:8888/User/username/Data/test_cubes/jw02732-o004_t004_miri_ch1-shortmediumlong_s3d.fits) with scheme  is not currently supported.
```

bmorris3 replied:
It looks like the string that was passed is a markdown link – what were your inputs?

rosteen replied:
I don't know where the markdown came from, but as we discussed offline this does work if I actually get the path to my file correct, I think we just need a slightly more informative error for the case where the filepath provided doesn't exist.

bmorris3 replied:
We solved this offline – the file path did not exist locally, but the error message wasn't informative. I put a more informative error message in fa83f48.

rosteen commented on jdaviz/utils.py:
```diff
@@ -385,3 +390,72 @@ def total_masked_first_data(self):
+        return download_file(possible_uri, cache=cache)
+
+    elif parsed_uri.scheme == '':
+        raise ValueError(f"The data to load '{possible_uri}' cannot be parse as a "
+                         f"URL or URI, and no existing local file is available "
+                         f"at this path.")
```
```suggestion
        raise ValueError(f"The input file '{possible_uri}' cannot be parsed as a "
                         f"URL or URI, and no existing local file is available "
                         f"at this path.")
```

parse -> parsed and another slight change.

rosteen commented on docs/imviz/import_data.rst:
```diff
@@ -139,6 +139,34 @@ data entries into the viewer until after the parsing is complete::
+
+The examples above import data from a local file path, and also support loading remote
+data from a URL or URI with :meth:`~jdaviz.core.helpers.ConfigHelper.load_data`.
+URL. If the input is a string with a MAST URI, the file will be retrieved via
```
```suggestion
If the input is a string with a MAST URI, the file will be retrieved via
```
Looks like this "URL." was leftover from a previous edit.

rosteen reviewed (commented):
I left one more suggestion and a comment on the caching default debate.

pllim commented on docs/conf.py:
```diff
@@ -263,6 +263,7 @@
 # Extra intersphinx in addition to what is already in sphinx-astropy
 intersphinx_mapping.update({  # noqa: F405
     'glueviz': ('https://docs.glueviz.org/en/stable/', None),
+    'astroquery': ('https://astroquery.readthedocs.io/en/stable/', None),
```
Given upstream service provider stuff can change faster than release, would `latest` be safer here? I dunno.

bmorris3 replied:
I don't know either, I chose stable since all other intersphinx links were stable.

pllim replied:
glue points to latest 🤷‍♀️

pllim commented on docs/imviz/import_data.rst:
```diff
@@ -139,6 +139,34 @@ data entries into the viewer until after the parsing is complete::
+``cache`` argument, which will store the file locally. Cached downloads via astropy
+are placed in the `astropy cache <https://docs.astropy.org/en/stable/utils/data.html>`_,
+and URIs retrieved via astroquery can be saved to a path of your choice with
+``local_path``::
```
To be consistent with your past doc edits.

```suggestion
``local_path``.

.. code-block:: python
```

pllim commented on docs/imviz/import_data.rst:
Should also discuss when the cache would default to True or False, and what is the implication of that, in a way that most users can understand.

pllim commented on jdaviz/configs/cubeviz/plugins/tests/test_parsers.py:
```diff
@@ -213,7 +213,7 @@ def test_numpy_cube(cubeviz_helper):
 
 def test_invalid_data_types(cubeviz_helper):
-    with pytest.raises(FileNotFoundError, match='Could not locate file'):
+    with pytest.raises(ValueError, match='URI does_not_exist.fits'):
```
```suggestion
    with pytest.raises(ValueError, match=r'URI does_not_exist\.fits'):
```

pllim commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -259,7 +259,7 @@ def mos_spec1d_parser(app, data_obj, data_labels=None,
 @data_parser_registry("mosviz-spec2d-parser")
 def mos_spec2d_parser(app, data_obj, data_labels=None, add_to_table=True,
-                      show_in_viewer=False, ext=1, transpose=False):
+                      show_in_viewer=False, ext=1, transpose=False, cache=True):
```
Why does the default to True, unlike others? Maybe should clarify in Mosviz import data section you added?

pllim commented on jdaviz/tests/test_utils.py:
For all the tests, we should explicitly set `cache=True` to avoid unnecessary downloads.

havok2063 reviewed (commented):
Took a first pass at this.  In general it looks good to me.  All the examples in the PR work for me.  A few issues came up for me though.  

1. For both Specviz and Specviz2d, I tried `load_data` with `cache=True` and got the ` unexpected keyword argument 'cache'` error.  Imviz and Cubeviz worked fine when setting the cache keyword.  
2. I tried a proprietary MAST uri, e.g `uri = "mast:JWST/product/jw01759-o010_t009_miri_ch1-shortmediumlong_x1d.fits"`.  Both with and without being properly authenticated.  It failed to load both times with `FileNotFoundError: No such file: /Users/bcherinka/Work/notebooks/prototype_spec1d2d/jw01759-o010_t009_miri_ch1-shortmediumlong_x1d.fits
`.   I think it should fail with a message something like "User not authenticated, cannot download file.", if I'm not authenticated.  If there's a message output from astroquery, we could just reuse that.  After I authenticated I expected it to work and load the spectrum, as the public uri did.     

I would also prefer a default of `cache=True` but I also don't have a strong preference, as long as we document it clearly somewhere.

pllim commented:
@bmorris3 , I am still waiting for the rest of doc update (and the response to Brian C's review). FYI.

bmorris3 commented:
As discussed in tag up, in ac3b4d2 I've updated all of the defaults to `cache=None`. If `cache=None`, the file _is_ cached and a warning is raised saying roughly: "this file will be cached – if you don't want to see this message again, set cache explicitly". 

RE @havok2063 – I've fixed the missing connections in specviz and specviz2d, thanks for that!

bmorris3 commented:
@havok2063 – I've collected the response from `Observations.download_file`, and raise any messages if they're returned in 4fe0dbd: 

<img width="773" alt="Screen Shot 2024-05-24 at 11 46 50" src="https://github.com/spacetelescope/jdaviz/assets/3497584/7dd7e044-9bb7-4991-bd2b-596058441a97">

bmorris3 commented on jdaviz/utils.py:
```diff
@@ -385,3 +390,72 @@ def total_masked_first_data(self):
+    possible_uri : str or other
+        This input will be returned without changes if it is not a string,
+        or if it is a local file path to an existing file. Otherwise,
+        it will be parsed as a URI. Local file URIs beginning with ``file://``
+        are not supported by this method – nor are they necessary, since string
+        paths without the scheme work fine! Cloud FITS are not yet supported.
```
Good idea, noted in b8d399a.

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,102 @@ def total_masked_first_data(self):
+
+    elif parsed_uri.scheme.lower() in ['http', 'https', 'ftp']:
+        if cache_warning:
+            warnings.warn(warning_msg)
```
```suggestion
            warnings.warn(warning_msg, UserWarning)
```

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,102 @@ def total_masked_first_data(self):
+                         f"at this path.")
+
+    else:
+        raise ValueError(f"URI {possible_uri} with scheme {parsed_uri.scheme} is not "
```
Is it possible to get this logic covered?

pllim commented on notebooks/CubevizExample.ipynb:
```diff
@@ -55,11 +51,12 @@
-    "and provide the directory path on your system where you would like the file stored.\n",
+    "Now we load observations. If you already have the files on your local machine, you can \n",
+    "load them by passing their file paths to `load_data` as strings. For this example, \n",
+    "we will retrieve remote data from [MAST](https://masttest.stsci.edu/) via `astroquery`\n",
```
Is it time to remove the "test", @havok2063 ?
```suggestion
    "we will retrieve remote data from [MAST](https://mast.stsci.edu/) via `astroquery`\n",
```

havok2063 replied:
Yes I think it's safe to remove the test here.

pllim commented on notebooks/ImvizExample.ipynb:
```diff
@@ -83,17 +81,12 @@
-    "force it to re-download the file if desired.\n",
+    "Now we load observations. If you already have the files on your local machine, you can \n",
+    "load them by passing their file paths to `load_data` as strings. For this example, \n",
+    "we will retrieve remote data from [MAST](https://masttest.stsci.edu/) via `astroquery`\n",
```
```suggestion
    "we will retrieve remote data from [MAST](https://mast.stsci.edu/) via `astroquery`\n",
```

pllim commented on notebooks/Specviz2dExample.ipynb:
```diff
@@ -58,12 +54,12 @@
-    "the directory path on your system where you would like the file stored.\n",
+    "Now we load observations. If you already have the files on your local machine, you can \n",
+    "load them by passing their file paths to `load_data` as strings. For this example, \n",
+    "we will retrieve remote data from [MAST](https://masttest.stsci.edu/) via `astroquery`\n",
```
```suggestion
    "we will retrieve remote data from [MAST](https://mast.stsci.edu/) via `astroquery`\n",
```

pllim commented on notebooks/SpecvizExample.ipynb:
```diff
@@ -74,17 +72,12 @@
-    "force it to re-download the file if desired."
+    "Now we load observations. If you already have the files on your local machine, you can \n",
+    "load them by passing their file paths to `load_data` as strings. For this example, \n",
+    "we will retrieve remote data from [MAST](https://masttest.stsci.edu/) via `astroquery`\n",
```
```suggestion
    "we will retrieve remote data from [MAST](https://mast.stsci.edu/) via `astroquery`\n",
```

pllim reviewed (commented):
How do I access `local_path` from public `load_data` API? I would like that in the example notebooks because locally I already have a place for all the MAST downloads (not just for Jdaviz) that isn't the working directory and I want to reuse the cache there.

pllim commented on jdaviz/tests/test_utils.py:
Also please make sure that your tests aren't saving the downloaded files in the source directory. Run these tests locally and do a `git status` and make sure there are no new untracked files after. Thanks.

havok2063 commented:
I'm out on vacation now for a while and can't properly review.  It looks good on quick glance.  Feel free to merge when y'all feel it's ready. If I find any new issues I can file a new tickets.

bmorris3 commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +17,43 @@ def test_alpha_index_exceptions():
+
+@pytest.mark.remote_data
+def test_url_to_download_imviz(imviz_helper):
+    url = "https://www.astropy.org/astropy-data/tutorials/FITS-images/HorseHead.fits"
```
Got it. Thanks!

pllim commented on jdaviz/configs/cubeviz/plugins/parsers.py:
```diff
@@ -38,6 +39,13 @@ def parse_data(app, file_obj, data_type=None, data_label=None, parent=None):
+        Cache the downloaded file if the data are retrieved by a query
+        to a URL or URI.
+    local_path : str, optional
+        Cache remote files to this path.
```
```suggestion
        Cache remote files to this path. This is only used if data is requested from `astroquery.mast`.
```

pllim commented on jdaviz/configs/cubeviz/plugins/parsers.py:
```diff
@@ -38,6 +39,13 @@ def parse_data(app, file_obj, data_type=None, data_label=None, parent=None):
+    cache : None, bool, or str
+        Cache the downloaded file if the data are retrieved by a query
+        to a URL or URI.
+    local_path : str, optional
```
I wonder if we want to use a more generic `**kwargs` to future-proof the API but I guess this is okay for now.

bmorris3 replied:
I know you know this but while it's definitely future-proof, it's also non-specific and can cause user heartache. I don't take a specific stance, other than agreeing that 
> this is okay for now

pllim replied:
I should have clarified: I specifically meant in a future where we start to support Cloud FITS etc, and need to introduce a bunch of new keywords that only apply under certain conditions, like `local_path`.

That said... I think we already need to add `timeout`.

pllim commented on jdaviz/configs/imviz/plugins/parsers.py:
```diff
@@ -60,11 +61,24 @@ def parse_data(app, file_obj, ext=None, data_label=None, parent=None):
+        to a URL or URI.
+
+    local_path : str, optional
+        Cache remote files to this path.
```
```suggestion
        Cache remote files to this path. This is only used if data is requested from `astroquery.mast`.
```

pllim commented on jdaviz/configs/imviz/tests/test_parser.py:
```diff
@@ -234,15 +235,15 @@ def test_parse_asdf_in_fits_4d(self, imviz_helper, tmp_path):
     def test_parse_jwst_nircam_level2(self, imviz_helper):
-        # TODO: Change back to smaller number (30?) when ITSD is convinced it is them and not us.
-        #       Help desk ticket INC0183598, J. Quick.
-        filename = download_file(self.jwst_asdf_url_1, cache=True, timeout=100)
```
This makes me wonder... Do we also need to expose the `timeout` keyword to user API? I think it is an accepted keyword in both `astropy` and `astroquery`.

rosteen replied:
I think it would be good to expose it.

pllim commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -281,6 +282,11 @@ def mos_spec2d_parser(app, data_obj, data_labels=None, add_to_table=True,
+        Cache the downloaded file if the data are retrieved by a query
+        to a URL or URI.
+    local_path : str, optional
+        Cache remote files to this path.
```
```suggestion
        Cache remote files to this path. This is only used if data is requested from `astroquery.mast`.
```

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -60,13 +60,19 @@ def load_data(self, data, data_label=None, format=None, show_in_viewer=True,
+            Cache the downloaded file if the data are retrieved by a query
+            to a URL or URI.
+        local_path : str, optional
+            Cache remote files to this path.
```
```suggestion
            Cache remote files to this path. This is only used if data is requested from `astroquery.mast`.
```

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -34,6 +35,11 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+        Cache the downloaded file if the data are retrieved by a query
+        to a URL or URI.
+    local_path : str, optional
+        Cache remote files to this path.
```
```suggestion
        Cache remote files to this path. This is only used if data is requested from `astroquery.mast`.
```

pllim commented on jdaviz/configs/specviz2d/helper.py:
```diff
@@ -63,6 +63,13 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
+            to a URL or URI.
+
+        local_path : str, optional
+            Cache remote files to this path.
```
```suggestion
            Cache remote files to this path. This is only used if data is requested from `astroquery.mast`.
```

pllim commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +16,23 @@ def test_alpha_index_exceptions():
+
+def test_uri_to_download_bad_scheme(imviz_helper):
+    uri = "file://path/to/file.fits"
+    with pytest.raises(ValueError, match='URI file://path/to/file.fits with scheme file'):
```
```suggestion
    with pytest.raises(ValueError, match=r'URI file://path/to/file\.fits with scheme file'):
```

pllim commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +16,23 @@ def test_alpha_index_exceptions():
+
+
+@pytest.mark.remote_data
+def test_uri_to_download_specviz(specviz_helper):
```
Please utilize your `local_path` so the file does not download to source dir.
```suggestion
def test_uri_to_download_specviz(specviz_helper, tmp_path):
```

pllim commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +16,23 @@ def test_alpha_index_exceptions():
+
+
+@pytest.mark.remote_data
+def test_uri_to_download_specviz2d(specviz2d_helper):
```
Same comment as above about `local_path`.
```suggestion
def test_uri_to_download_specviz2d(specviz2d_helper, tmp_path):
```

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,117 @@ def total_masked_first_data(self):
         return cls(masks=masks)
+
+
+def download_uri_to_path(possible_uri, cache=None, local_path=os.curdir):
```
Since the mention of standalone app this morning, I wonder now if `os.curdir` is okay in standalone mode. Don't we want something that is consistent with Export plugin behavior?

bmorris3 replied:
My take is: if you've run standalone from your terminal, you might expect that files get saved to the path where you launched jdaviz.

pllim replied:
Right, that is the expectation, but working dir is actually some `/tmp` dir the voila server is running from. You would need something like this to access the path where you launched Jdaviz via CLI.

https://github.com/spacetelescope/jdaviz/blob/50ebc1ebbfe24f9e3e5c01c87d8e7e0d3772a36d/jdaviz/configs/default/plugins/export/export.py#L349-L350

bmorris3 replied:
Ah, I didn't realize this, thanks. (and don't use standalone enough to have found this out naturally 🙃)

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,117 @@ def total_masked_first_data(self):
+    local_path : str, optional
+        Save the downloaded file to this path. Default is to
+        save the file with its remote filename in the current
+        working directory.
```
```suggestion
        working directory. This is only used if data is requested from `astroquery.mast`.
```

pllim reviewed (commented):
p.s. `local_path` might trip up some users because it does not really accept a directory path, just a filename path. I wonder if we have to guard against https://github.com/astropy/astroquery/issues/2501 but then again, this is not really our problem.

p.p.s. If the single backtick for `astroquery.mast` gives warning, please use double backtick to avoid intersphinx.

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,117 @@ def total_masked_first_data(self):
+
+        if local_path is None:
+            # if not specified, this is the default location:
+            local_path = os.path.join(os.getcwd(), parsed_uri.path.split(os.path.sep)[-1])
```
Same comment about standalone app path above.

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,117 @@ def total_masked_first_data(self):
+            warnings.warn(cache_none_msg, UserWarning)
+
+        if local_path not in (os.curdir, None):
+            warnings.warn(local_path_msg, UserWarning)
```
This line isn't covered by test?

bmorris3 replied:
Is now 😃

pllim reviewed (commented):
Ooops, I submitted review too early. Here is the rest of it.

FWIW, I like the clean-up in the notebooks from this new feature. 👏

rosteen reviewed (commented):
It worked in my testing, as said in a comment I really like how it cleans up the example code. I'll be happy to approve after you consider my suggestion and address Pey Lian's comments (I think I agree that it would be good to expose `timeout`, which I hope isn't a very big effort).

bmorris3 commented on notebooks/CubevizExample.ipynb:
```diff
@@ -74,15 +71,11 @@
     "with warnings.catch_warnings():\n",
     "    warnings.simplefilter('ignore')\n",
-    "    cubeviz.load_data(f'{data_dir}/{fn}')"
+    "    cubeviz.load_data(uri, cache=True)"
```
Love to hear it.

bmorris3 commented on notebooks/CubevizExample.ipynb:
```diff
@@ -55,11 +51,12 @@
+    "load them by passing their file paths to `load_data` as strings. For this example, \n",
+    "we will retrieve remote data from [MAST](https://mast.stsci.edu/) via `astroquery`\n",
+    "by passing the observation's URI as a string. By default, the downloaded files are \n",
+    "saved to the current working directory. If you set `cache=False` in the `load_data` \n",
+    "call, it will re-download the file if desired.\n",
```
Counter-intuitively (to my taste), `cache=True` means load an existing file if there is one, and if there isn't, store it; whereas `cache=False` means don't load existing files even if they exist, and save the downloaded copy.

pllim replied:
Are you talking about astropy or astroquery?

bmorris3 replied:
astroquery. And astropy is different. 🙈

pllim commented on jdaviz/configs/cubeviz/plugins/parsers.py:
```diff
@@ -38,6 +39,18 @@ def parse_data(app, file_obj, data_type=None, data_label=None, parent=None):
+    timeout : float, optional
+        If downloading from a remote URL, set the timeout limit for
+        remote requests in seconds (passed to
+        `~astropy.utils.data.download_file`).
```
astroquery also takes this, theoretically, unless you have checked that it doesn't?

pllim replied:
OK but that isn't reflected in the docstring here. Only `astropy.utils.data.download_file` is mentioned.

pllim commented on jdaviz/configs/specviz2d/helper.py:
```diff
@@ -63,6 +63,14 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
+
+        local_path : str, optional
+            Cache remote files to this path. This is only used if data is
+            requested from `astroquery.mast`.
```
timeout doc missing here.

pllim commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +17,41 @@ def test_alpha_index_exceptions():
+@pytest.mark.remote_data
+def test_uri_to_download_specviz(specviz_helper, tmp_path):
+    uri = "mast:JWST/product/jw02732-o004_t004_miri_ch1-shortmediumlong_x1d.fits"
+    specviz_helper.load_data(uri, cache=True)
```
I think you forgot to do the `local_path = str(tmp_path / blahblah)` stuff here.

pllim commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +17,41 @@ def test_alpha_index_exceptions():
+    uri = "mast:JWST/product/jw01324-o006_s00005_nirspec_f100lp-g140h_s2d.fits"
+    with warnings.catch_warnings():
+        warnings.simplefilter('ignore', AsdfWarning)
+        specviz2d_helper.load_data(uri, cache=True)
```
Same comment as https://github.com/spacetelescope/jdaviz/pull/2875/files#r1624551389

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,127 @@ def total_masked_first_data(self):
+        local path to the downloaded file.
+    """
+
+    if local_path == os.curdir and os.environ.get("JDAVIZ_START_DIR", ""):
```
I don't think it matters what the `local_path` is when in this mode? You would want to grab the JDAVIZ_START_DIR regardless.
```suggestion
    if os.environ.get("JDAVIZ_START_DIR", ""):
```

bmorris3 replied:
Ok, got it.

pllim commented on jdaviz/tests/test_utils.py:
```diff
@@ -14,3 +17,41 @@ def test_alpha_index_exceptions():
+
+@pytest.mark.remote_data
+def test_url_to_download_imviz_local_path_warning(imviz_helper):
+    url = "https://www.astropy.org/astropy-data/tutorials/FITS-images/HorseHead.fits"
+    with (
+        pytest.warns(UserWarning, match='You requested to cache data'),
+        pytest.warns(FITSFixedWarning, match="'datfix' made the change")
+    ):
+        imviz_helper.load_data(url, cache=False, local_path='horsehead.fits')
```
I don't quite understand the expected behavior of this test. If I give a direct URL, I would expect the warning to be about `local_path` being ignored because it is going through `astropy.utils.data.download_file` and not `astroquery.mast`. Instead here you are checking that the expected warning to be about `cache`. Why?

bmorris3 replied:
That's actually what the warning is: 
```
    local_path_msg = (
        f"You requested to cache data to the `local_path`='{local_path}'. This "
        f"keyword argument is supported for downloads of MAST URIs via astroquery, "
        f"but since the remote file at '{possible_uri}' will be downloaded "
        f"using `astropy.utils.data.download_file`, the file will be "
        f"stored in the astropy download cache instead."
    )
```

bmorris3 commented on jdaviz/configs/specviz2d/helper.py:
```diff
@@ -132,6 +141,7 @@ def load_data(self, spectrum_2d=None, spectrum_1d=None, spectrum_1d_label=None,
                 spectrum_1d, data_label=spectrum_1d_label,
                 parser_reference="specviz-spectrum1d-parser",
                 show_in_viewer=show_in_viewer,
+                cache=cache
```
Yikes, good catch.

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -60,13 +60,20 @@ def load_data(self, data, data_label=None, format=None, show_in_viewer=True,
+            to a URL or URI.
+        local_path : str, optional
+            Cache remote files to this path. This is only used if data is
+            requested from `astroquery.mast`.
```
```suggestion
            requested from `astroquery.mast`.
        timeout : float, optional
            If downloading from a remote URI, set the timeout limit for
            remote requests in seconds (passed to
            `~astropy.utils.data.download_file` or
            `~astroquery.mast.Conf.timeout`).
```

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -60,13 +60,20 @@ def load_data(self, data, data_label=None, format=None, show_in_viewer=True,
                           show_in_viewer=show_in_viewer,
-                          concat_by_file=concat_by_file)
+                          concat_by_file=concat_by_file,
+                          cache=cache)
```
```suggestion
                          cache=cache, local_path=local_path, timeout=timeout)
```

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -9,15 +10,15 @@
 @data_parser_registry("specviz-spectrum1d-parser")
 def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_viewer=True,
-                              concat_by_file=False):
+                              concat_by_file=False, cache=None, local_path=os.curdir, timeout=None):
```
Why is this default different from the others?
```suggestion
                              concat_by_file=False, cache=None, local_path=None, timeout=None):
```

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,132 @@ def total_masked_first_data(self):
+        local path to the downloaded file.
+    """
+
+    if os.environ.get("JDAVIZ_START_DIR", ""):
+        # avoiding creating local paths in a tmp dir when in standalone:
+        local_path = os.environ["JDAVIZ_START_DIR"] / local_path
+
+    if not isinstance(possible_uri, str):
+        # only try to parse strings:
+        return possible_uri
+
+    if os.path.exists(possible_uri):
+        # don't try to parse file paths:
+        return possible_uri
```
The diff in suggestion might look crazy but I am merely moving the `JDAVIZ_START_DIR` check to after the `return` checks. There is no reason to overwrite `local_path` before those checks.
```suggestion
    if not isinstance(possible_uri, str):
        # only try to parse strings:
        return possible_uri

    if os.path.exists(possible_uri):
        # don't try to parse file paths:
        return possible_uri

    if os.environ.get("JDAVIZ_START_DIR", ""):
        # avoiding creating local paths in a tmp dir when in standalone:
        local_path = os.environ["JDAVIZ_START_DIR"] / local_path
```

pllim commented on jdaviz/utils.py:
```diff
@@ -1,20 +1,26 @@
 from astropy.utils import minversion
+from astropy.utils.data import download_file
 from astropy.wcs.wcsapi import BaseHighLevelWCS
+from astroquery.mast import Observations, Conf
```
I think you actually want the lowercase `conf` (instantiated mast config). See https://github.com/astropy/astroquery/blob/25e6521ad3e86f5b2285d08e736cd3417159482d/astroquery/mast/__init__.py#L31

```suggestion
from astroquery.mast import Observations, conf
```

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,132 @@ def total_masked_first_data(self):
+            local_path = os.path.join(local_path, parsed_uri.path.split(os.path.sep)[-1])
+
+        if timeout is None:
+            timeout = Conf.timeout.defaultvalue
```
```suggestion
            timeout = conf.timeout.defaultvalue
```

pllim commented on jdaviz/utils.py:
```diff
@@ -385,3 +391,132 @@ def total_masked_first_data(self):
+        if timeout is None:
+            timeout = Conf.timeout.defaultvalue
+
+        with Conf.timeout.set_temp(timeout):
```
```suggestion
        with conf.timeout.set_temp(timeout):
```

bmorris3 replied:
Thanks, just caught these too in 77df5ff.

pllim replied:
If we want to be more pedantic, we could `from astroquery.mast import conf as mast_conf` but since there isn't any name clash now, doesn't matter.

pllim reviewed (approved):
LGTM now. Thanks!

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -1,3 +1,4 @@
+import os
```
flake8 check failed.
```suggestion
```

pllim commented:
Hmm I think the test failures are real: https://github.com/spacetelescope/jdaviz/actions/runs/9353171515/job/25743015163?pr=2875

pllim commented:
Ah, by changing from `Conf` to `conf`, the syntax to access default value might be different now. See the last example in https://docs.astropy.org/en/stable/config/index.html#exploring-configuration

bmorris3 commented:
Remaining failures are unrelated. Thanks @rosteen and @pllim!
